pixeltable/utils/arrow.py

```diff
@@ -0,0 +1,112 @@
+    elif isinstance(pixeltable_type, ts.ArrayType):
+        return pa.fixed_shape_tensor(pa.from_numpy_dtype(pixeltable_type.numpy_dtype()), pixeltable_type.shape)
+    else:
+        assert False, f'default conversion for {pixeltable_type} not implemented'
+
+def get_pixeltable_schema(arrow_schema : pa.Schema) -> Dict[str, ts.ColumnType]:
+    return {field.name: get_pixeltable_type(field.type) for field in arrow_schema}
+
+def get_arrow_schema(pixeltable_schema : Dict[str, Any]) -> pa.Schema:
+    return pa.schema({name: get_arrow_type(dtype) for name, dtype in pixeltable_schema.items()})
+
+def to_pydict(batch : pa.RecordBatch) -> Dict[str, Iterable[Any]]:
+    """ Convert a RecordBatch to a dictionary of lists, unlike pa.lib.RecordBatch.to_pydict,
+        this function will not convert numpy arrays to lists, and will preserve the original numpy dtype.
+    """
+    out = {}
+    for k, name in enumerate(batch.schema.names):
+        col = batch.column(k)
+        if isinstance(col.type, pa.FixedShapeTensorType):
+            # treat array columns as numpy arrays to easily preserve numpy type
+            out[name] = col.to_numpy(zero_copy_only=False)
+        else:
+            # for the rest, use pydict to preserve python types
+            out[name] = col.to_pylist()
+
+    return out
+
+def iter_tuples(batch : pa.RecordBatch) -> Iterator[Dict[str,Any]]:
+    """ Convert a RecordBatch to an iterator of dictionaries. also works with pa.Table and pa.RowGroup """
+    pydict = to_pydict(batch)
+    n = len(pydict[list(pydict.keys())[0]])
+    for i in range(n):
+        yield {col_name: values[i] for col_name, values in pydict.items()}
+
+
+def create_table_from_parquet(cl : 'pixeltable.Client',
+                              path_str: str,
+                              parquet_path: str,
+                              explain_only : bool = False,
+                              primary_key: Union[str, List[str]] = [],
+                              num_retained_versions: int = 10) -> Optional['catalog.InsertableTable']:
+    """ See pixeltable.Client.create_table_from_parquet for documentation
+    """
+
+    input_path = pathlib.Path(parquet_path).expanduser()
+    parquet_dataset = pa.parquet.ParquetDataset(input_path)
+    arrow_schema = parquet_dataset.schema
+    base_schema = get_pixeltable_schema(arrow_schema)
+
+    print('mapping parquet columns to pixeltable schema:', base_schema)
+    if explain_only:
+        return None
+
+    tab = cl.create_table(path_str, base_schema, primary_key, num_retained_versions)
+
+    try:
+        for fragment in parquet_dataset.fragments:
+            for batch in fragment.to_batches():
+                dict_batch = [d for d in iter_tuples(batch)]
+                tab.insert(dict_batch)
+    except Exception as e:
+        _logger.error(f'Error while inserting Parquet file into table: {e}')
+        cl.drop_table(path_str, force=True, ignore_errors=True)
+        raise e
```

Comment:
I don't love this pattern. Better would be a transaction pattern (does the pixeltable Client support this? If not, something to discuss with @mkornacker). [What if something goes wrong with the call to `cl.drop_table`?]

Reply:
Yeah, it's not ideal, but there's no transaction api. this feels like it would cover a big fraction of cleanups from users, eg you interrupted your notebook mid way). Users can clean up any exceptions to this otherwise. 

Maybe we can do something like table with temporary name, then rename to user specified name if everything is good.

Reply:
That's probably better. Even just leaving the partial table in place wouldn't be terrible, as long as we're clear about surfacing the exception to the user. (That's not an uncommon way for systems to react to a broken data transfer. The user can always choose to delete the partial table, and the partial output might be useful for debugging.)